### DISCUSSION TITLE
Created a face that I'd like to contribute

### DISCUSSION THREAD

**Discussion by BrianBinFL** [joeycastillo/Sensor-Watch#300] [Q&A] [answered]
I had the impulse to create a minute repeater function for the Sensor Watch and was pleased to find that Jonas Termeau had already written repetition_minute_face. It's in the repo and available in the Sensor Watch Build tool. Thanks to Jonas for writing it.

The implementation that Jonas wrote uses the more traditional hours, quarters, and minutes pattern. But for purposes of ease of use the decimal minute repeater pattern of hours, tens, and minutes is much more friendly. I copied Jonas' face, and named the copy minute_repeater_decimal_face. I changed it to a decimal minute repeater and added in a 500ms delay at the end of the hours and again at the end of the tens, to add a little gap of time between the groupings of tones.

I'd like to contribute my modified face to the repo, and have it be available in the Sensor Watch Build tool. But sadly I'm VERY ignorant about using the source code repository to do so. I figured out how to install the various GIT applications on my Ubuntu build, authenticate the GIT CLI to GitHub, clone the repo to my Ubuntu machine (that I created solely for this purpose), etc.

I then copied the repetition_minute_face.c and repetition_minute_face.h files to minute_repeater_decimal_face.c and minute_repeater_decimal_face.h respectively. Of course I had to edit the various include files to announce the existence of the new face. And I had to modify some of the function names in my new .C and .H files so they wouldn't duplicate the ones in the original repetition_minute_face.

It all compiles and works fine. I have a build running on my watch that includes my modified minute_repeater_decimal_face.

But in all of the above I didn't create any branches on GitHub, nor do I really know how to, or anything else. I know that ultimately I need to create a Pull Request. But I think there has to be a branch in order to do that. And honestly I'm just parroting things that I've read at this point.

I know it's asking a lot, but if someone could tell me what I need to do in order to contribute the new face I'd be most appreciative. I think I learn pretty quick, and I have no problem searching and reading to learn how to do things. But I don't want to do anything wrong or make a mess of anything. So if someone who knows what they are doing can guide me that would be wonderful.

Thanks!

**Comment by BrianBinFL** [comment]
I realize that my request was way too much to ask. There's just too much to the whole git and GitHub universe. So I purchased some git and GitHub courses online and am getting up to speed. Maybe in a week or so I'll be ready to take a stab at a pull request for the face I created.

**Comment by wryun** [comment, marked as answer]
One thing I'd say is don't worry about:

> But I don't want to do anything wrong or make a mess of anything.

You don't have permissions to mess things up. The way it will work, briefly, is:

- you'll create a fork of the repository in github
- you'll clone that fork locally
- create a new branch locally
- make your change locally on that branch
- commit that change
- push that branch back to your fork on github
- create a 'Pull Request' on _this_ repository referencing the branch in your fork

The only thing that affects this project at all is the final step, and it's easy to just close it if it's wrong. So feel free to experiment and make mistakes!

By the way, the Discord seems to be slightly more active than here (sadly :laughing: ), so if you get stuck you might get more of a response there.

**Reply by BrianBinFL** [reply]
This is actually exactly what I needed. I finished my first Git and GitHub class and learned most of the things that you mention in your sequence of operations. I'm in the middle of doing a more advanced class now which will cover pull requests, etc.

I tried to take a look at the discord at one point but it seems I set up a discord account at one point and I need to untangle something with it. I'll get that sorted when I have a moment.

Again, thanks for the above, that list of steps is exactly the blanks I needed filled in.

**Reply by BrianBinFL** [reply]
Having completed my Git and GitHub classes I understand your instructions very well at this point. A question about a fine point if I may.

I think I would add a couple steps to those that you provided - but I want to make sure this doesn't create a problem. New steps are in bold.

1. you'll create a fork of the repository in github
2. you'll clone that fork locally
3. create a new branch locally
4. make your change locally on that branch
5. commit that change
6. **merge my new local branch into my local main**
7. **edit the movement_config.h file to include my new face**
8. **compile everything**
9. **copy the watch.uf2 file to the watch**
10. **make sure everything is happy**
11. push that branch back to your fork on github
12. create a 'Pull Request' on this repository referencing the branch in your fork

When I create my pull request as long as I specify the **branch** that I created and not my main branch, I think all of the changes to movement_config.h and any of the files created from compiling won't be included in the pull request since they weren't changes made as part of the branch.

Hopefully I got all that right.

**Reply by BrianBinFL** [reply]
Well, I added PR number 308 and best I can tell from the diff display it looks like I did it right. Fingers crossed. :-)
